neo/SmartContract/Native/Tokens/NeoToken.cs

```diff
@@ -0,0 +1,291 @@
+                {
+                    amount += (Blockchain.DecrementInterval - istart) * Blockchain.GenerationAmount[ustart];
+                    ustart++;
+                    istart = 0;
+                }
+                amount += (iend - istart) * Blockchain.GenerationAmount[ustart];
+            }
+            amount += (uint)(engine.Service.Snapshot.GetSysFeeAmount(end - 1) - (start == 0 ? 0 : engine.Service.Snapshot.GetSysFeeAmount(start - 1)));
+            return value * amount * GAS.Factor / TotalAmount;
+        }
+
+        private bool Initialize(ApplicationEngine engine, UInt160 account)
+        {
+            if (engine.Service.Trigger != TriggerType.Application) throw new InvalidOperationException();
+            if (base.TotalSupply(engine) != BigInteger.Zero) return false;
+            MintTokens(engine, account, TotalAmount);
+            foreach (ECPoint pubkey in Blockchain.StandbyValidators)
+                RegisterValidator(engine, pubkey.EncodePoint(true));
+            return true;
+        }
+
+        private BigInteger UnclaimedGas(ApplicationEngine engine, UInt160 account, uint end)
+        {
+            StorageItem storage = engine.Service.Snapshot.Storages.TryGet(CreateAccountKey(account));
+            if (storage is null) return BigInteger.Zero;
+            AccountState state = new AccountState(storage.Value);
+            return CalculateBonus(engine, state.Balance, state.BalanceHeight, end);
+        }
+
+        private bool RegisterValidator(ApplicationEngine engine, byte[] pubkey)
+        {
+            if (pubkey.Length != 33 || (pubkey[0] != 0x02 && pubkey[0] != 0x03))
+                throw new ArgumentException();
+            StorageKey key = CreateStorageKey(Prefix_Validator, pubkey);
+            if (engine.Service.Snapshot.Storages.TryGet(key) != null) return false;
+            engine.Service.Snapshot.Storages.Add(key, new StorageItem
+            {
+                Value = new ValidatorState().ToByteArray()
+            });
+            return true;
+        }
+
+        private bool Vote(ApplicationEngine engine, UInt160 account, ECPoint[] pubkeys)
+        {
+            if (!engine.Service.CheckWitness(engine, account)) return false;
+            StorageKey key_account = CreateAccountKey(account);
+            if (engine.Service.Snapshot.Storages.TryGet(key_account) is null) return false;
+            StorageItem storage_account = engine.Service.Snapshot.Storages.GetAndChange(key_account);
+            AccountState state_account = new AccountState(storage_account.Value);
+            foreach (ECPoint pubkey in state_account.Votes)
+            {
+                StorageItem storage_validator = engine.Service.Snapshot.Storages.GetAndChange(CreateStorageKey(Prefix_Validator, pubkey.ToArray()));
+                ValidatorState state_validator = ValidatorState.FromByteArray(storage_validator.Value);
+                state_validator.Votes -= state_account.Balance;
```

Comment:
Which balance is being subtracted here? It should be the previous balance before the last vote, no?

Reply:
Not balance, but votes.